### ISSUE TITLE
Incorrect Serializer behaviour for BooleanField(required=False) when passed a QueryDict

### ISSUE THREAD

**Issue by Florian-Projects** [encode/django-rest-framework#8300]
## Checklist

- [ ] Raised initially as discussion #...
- [x] This cannot be dealt with as a third party library. (We prefer new functionality to be [in the form of third party libraries](https://www.django-rest-framework.org/community/third-party-packages/#about-third-party-packages) where possible.)
- [x] I have reduced the issue to the simplest possible case.

The [Documentation](https://www.django-rest-framework.org/api-guide/fields/) states the following about the `required` keyword argument: 
"Normally an error will be raised if a field is not supplied during deserialization. Set to false if this field is not required to be present during deserialization.

Setting this to False also allows the object attribute or dictionary key to be omitted from output when serializing the instance. If the key is not present it will simply not be included in the output representation. "

However this doesn't work correctly with django.http.QueryDict. When passing a QueryDict to the Serializer, the not required fields are not omitted in the output but instead are set to False.

## Affected Version:
django_rest_framework <= 3.12
found when upgrading from 3.11 to latest version

## A Code Example
```python
from django.http import QueryDict
from rest_framework import serializers

data_dict = {'required': True}
data_querydict = QueryDict('required=True')


class TestSerializer(serializers.Serializer):
    required = serializers.BooleanField(required=True)
    param = serializers.BooleanField(required=False)


serializer1 = TestSerializer(data=data_dict)
serializer1.is_valid()

serializer2 = TestSerializer(data=data_querydict)
serializer2.is_valid()

print(f"Serialized Dict: {serializer1.validated_data}")
print(f"Serialized QueryDict: {serializer2.validated_data}")

if serializer1.validated_data != serializer2.validated_data:
    print("There is a Bug")
```
When executing this Example you will see that:
`serializer1.validated_data == OrderedDict([('required', True)])` while `serializer2.validated_data == OrderedDict([('required', True), ('param', False)])`. <br>
**This diverging behaviour is not present in other types of Field (Tested myself: CharField, IntegerField)**
This is potentially an API breaking change

## Possible Workaround
As Described [here](https://github.com/encode/django-rest-framework/issues/7582)
```python
from rest_framework import fields, serializers


class DefaultEmptyBooleanField(fields.BooleanField):
    default_empty_html = fields.empty


class MySerializer(serializers.Serializer):
    boolean_field = DefaultEmptyBooleanField(required=False)
```
## Expected Behavior
not required fields are omitted in the output for both Dict and QueryDict

**Comment by grazzolini** [comment]
I have this exact same issue, but with a URLField. In my scenario, the field is not present on the data, but given it has `allow_null=True`, `allow_blank=True` and _also_ `required=False`, the field, even if not present on the initial JSON data, is set to `None` on the serialized data, after `is_valid()` is called.

**Comment by grazzolini** [comment]
I think a few more checks should be done here: https://github.com/encode/django-rest-framework/blob/master/rest_framework/fields.py#L440-L447. I'll try to put a PR up to fix this.

**Comment by grazzolini** [comment]
This is not stale, this bug is still present on DRF.

**Comment by TonyMistark** [comment]
> 

```
from rest_framework.serializers import Serializer
from rest_framework.serializers import BooleanField
from django.http import QueryDict
class TestS(Serializer):
    bool_field_1 = BooleanField(required=False)
    bool_field_2 = BooleanField(required=False, default=False)
s = TestSerializer(data=QueryDict(), partial=True)
s.is_valid(raise_exception=True)
s.validated_data
# output
# OrderedDict()

s = TestSerializer(data={}, partial=True)
s.is_valid(raise_exception=True)
s.validated_data
# output
# OrderedDict()
```
bool_field_2 = BooleanField(required=False, default=False)
bool_field_2 had set default value
so
## Expected output should be:
OrderedDict([('bool_field_2', False)])

send partial=True, It will not solve the problem completely, but will create new problems

**Comment by TonyMistark** [comment]
> You can also use `partial` option for temporary purposes.
> 
> ```python
> serializer_partial = TestSerializer(data=data_querydict, partial=True)
> serializer_partial.is_valid()
> serializer_partial.validated_data
> ```
> 
> And the BooleanField's test cases show the reason: for HTML checkboxes.
> 
> https://github.com/encode/django-rest-framework/blob/f46c33e4e2c545e3fd90de692242c53a6fe6e7e9/tests/test_fields.py#L362-L385
> 
> Doc: https://www.django-rest-framework.org/api-guide/serializers/#partial-updates

https://github.com/encode/django-rest-framework/issues/8300#issuecomment-1293373562

**Comment by TonyMistark** [comment]
> You can also use `partial` option for temporary purposes.
> 
> ```python
> serializer_partial = TestSerializer(data=data_querydict, partial=True)
> serializer_partial.is_valid()
> serializer_partial.validated_data
> ```
> 
> And the BooleanField's test cases show the reason: for HTML checkboxes.
> 
> https://github.com/encode/django-rest-framework/blob/f46c33e4e2c545e3fd90de692242c53a6fe6e7e9/tests/test_fields.py#L362-L385
> 
> Doc: https://www.django-rest-framework.org/api-guide/serializers/#partial-updates


```
from rest_framework.serializers import Serializer
from rest_framework.serializers import BooleanField
from django.http import QueryDict
class TestS(Serializer):
    bool_field_1 = BooleanField(required=False)
    bool_field_2 = BooleanField(required=False, default=False)
s = TestSerializer(data=QueryDict(), partial=True)
s.is_valid(raise_exception=True)
s.validated_data
# output
# OrderedDict()

s = TestSerializer(data={}, partial=True)
s.is_valid(raise_exception=True)
s.validated_data
# output
# OrderedDict()
```
bool_field_2 = BooleanField(required=False, default=False)
bool_field_2 had set default value
so
## Expected output should be:
OrderedDict([('bool_field_2', False)])

send partial=True, It will not solve the problem completely, but will create new problems

**Comment by TonyMistark** [comment]
> Oh we have a same problem with `NullBooleanField`s:
> 
> ```python
> class TestSerializer(serializers.Serializer):
>     test = serializers.NullBooleanField(required=False)
> 
> bla = TestSerializer(data=QueryDict())
> bla.is_valid()
> bla.validated_data
> ```
> 
> `OrderedDict([('test', False)])`
> 
> Quick fix for us atm `TestSerializer(data=self.request.query_params.dict())`

`NullBooleanField` resolved my problom`
### Both types of empty dictionary input give the same result
```
from rest_framework.serializers import Serializer
from rest_framework.serializers import NullBooleanField
from django.http import QueryDict

class TestSerializer(Serializer):
    null_bool = NullBooleanField(required=False)
s = TestSerializer(data={})
s.is_valid(raise_exception=True)
print("input {}: ", s.validated_data)
# output
# OrderedDict()

s = TestSerializer(data=QueryDict())
s.is_valid(raise_exception=True)
print("input QueryDict: ", s.validated_data)
# output
# OrderedDict()
```
```
input {}:  OrderedDict()
input QueryDict:  OrderedDict()
```

**Comment by auvipy** [comment]
I am open for a suggested fix here. you can start with at least the failing test

**Comment by auvipy** [comment]
I would love your inputs and trying this PR https://github.com/encode/django-rest-framework/pull/8614. also please contribute some suggested test cases to the test suite of this project.

**Comment by tanushree05** [comment]
I want to contribute to this issue please give me permission

**Comment by merwok** [comment]
I don’t think you need permission, you can report if the changes in #8614 fix the problems and if not submit a pull request!

**Comment by trevnocap** [comment]
This issue is not stale, and still needs resolved

```python
from rest_framework.serializers import Serializer
from rest_framework.views import APIView

class TestAPI(APIView):
    
    class InputSerializer(serializers.Serializer):
        id = serializers.UUIDField()
        required_field1 = serializers.BooleanField(required=True)
        required_field2 = serializers.BooleanField(required=True)

    def put(self, request):
        serializer = self.InputSerializer(data=request.data)
        print(serializer.initial_data) 
        serializer.is_valid(raise_exception= True)
        print(serializer.validated_data)
```        

With `request.data` is `<QueryDict: {'id: ['uuid_value'], 'required_field1': ['true']}>`, I get `serializer.validated_data` prints: `OrderedDict([('division_setting_id', UUID('uuid_value')), ('require_field1', True), ('required_fields2', False)])`

**Comment by ArtemIsmagilov** [comment]
in the web api, I want to add the `first_id` and `second_id` parameters, but these parameters are not required in the web form, although in the model from which the serializer was created, these ForeigenKey parameters are required

**Comment by JIVAN-SUBEDI** [comment]
you can use partial option to fix the problem
serializer_partial = testserializer(data=data_querydict, partial=True)
serializer_partial.is_valid()
serializer_partial.validated_data

**Comment by p-r-a-v-i-n** [comment]
I have invest my time to investigate this. I went through each line of code of is_valid() function.

![Image](https://github.com/user-attachments/assets/9f445f8b-4297-4fce-8741-21e958f7f94b)

and what i have figureout that get_value function is what causing this behaviour.
when input is being sent query dict this preventing the it from defaulting to False.
```python
if self.field_name not in dictionary:
    if getattr(self.root, 'partial', False):
        return empty
    return self.default_empty_html # this where the problem is 
```
Simple fix i could consider here just to skip entire if not required.
for example
```python
if self.field_name not in dictionary:
    if getattr(self.root, 'partial', False):
        return empty
    if not self.required:
        return empty  # Skip this field entirely
    return self.default_empty_html
```

you could see the behaviour below

![Image](https://github.com/user-attachments/assets/c0090466-a547-4ce9-9bd4-e372597841a3)

I would love to open PR for this if it make sense.

**Comment by auvipy** [comment]
@browniebroke what do think about the latest proposed change?

**Comment by browniebroke** [comment]
Ok, I took some time to read the whole thread trying I understand the problem. The second part of [this message](https://github.com/encode/django-rest-framework/issues/8300#issuecomment-1034913616) was helpful to understand why we this behaviour:

> And the [BooleanField's test cases](https://github.com/encode/django-rest-framework/blob/f46c33e4e2c545e3fd90de692242c53a6fe6e7e9/tests/test_fields.py#L362-L385) show the reason: for HTML checkboxes.

I couldn't find the exact snippet from @p-r-a-v-i-n - but if you have a GitHub link to it

> and what i have figureout that get_value function is what causing this behaviour.
> when input is being sent query dict this preventing the it from defaulting to False.
> ```python
> if self.field_name not in dictionary:
>     if getattr(self.root, 'partial', False):
>         return empty
>     return self.default_empty_html # this where the problem is 
> ```

Maybe the the last bit `return self.default_empty_html` is pseudo-code? 

In the actual code I found, it uses `if html.is_html_input(dictionary)`, which -if you look at its implementation- explains the problem:

https://github.com/encode/django-rest-framework/blob/f148e4e2593e55c76c746c85422924c33ec435cc/rest_framework/utils/html.py#L9-L12

This runs and both asserts are true:

```python
q = QueryDict('')
assert hasattr(q, "getlist")
d = {}
assert not hasattr(d, "getlist")
```

So basically DRF relies on the type of the input data to be a `QueryDict` to detect that it was submitted from a HTML input, which I assume means the browsable API. 

If that's correct, I think fixing this issue would potentially break the HTML forms in browsable API, or at least the ones with checkboxes.

**Comment by p-r-a-v-i-n** [comment]
> I couldn't find the exact snippet from [@p-r-a-v-i-n](https://github.com/p-r-a-v-i-n) - but if you have a GitHub link to it
> 
> > and what i have figureout that get_value function is what causing this behaviour.
> > when input is being sent query dict this preventing the it from defaulting to False.
> > if self.field_name not in dictionary:
> >     if getattr(self.root, 'partial', False):
> >         return empty
> >     return self.default_empty_html # this where the problem is
@browniebroke please find snippet here https://github.com/encode/django-rest-framework/blob/4bb46c2949fc67a1f8e74c43776833d81df471ed/rest_framework/fields.py#L418

**Comment by p-r-a-v-i-n** [comment]
@browniebroke I think coding it out in real may give more clarity and  evidence , I will check and let you know but strongly believe it could be fixed.

**Comment by browniebroke** [comment]
> I think coding it out in real may give more clarity and evidence , I will check and let you know but strongly believe it could be fixed.

Sure, feel free to give it a try.

Taking a step back, think that most of the examples show the problem at the serializer level, but I don't show the view or how the request is being made. I don't fully understand how the bug manifests in people application, and what they are trying to achieve. Is it a problem you're hitting while using the browsable API? Is it that you client sends its payload with content-type: form-data instead of JSON? Something else?  

It would be nice to get clarity on that, just to make sure there isn't a better to solve/workaround it in user-land.

**Comment by p-r-a-v-i-n** [comment]
Hi @browniebroke , Sorry late response.
> > I think coding it out in real may give more clarity and evidence , I will check and let you know but strongly believe it could be fixed.
> 
> Sure, feel free to give it a try.
> 
> Taking a step back, think that most of the examples show the problem at the serializer level, but I don't show the view or how the request is being made. I don't fully understand how the bug manifests in people application, and what they are trying to achieve. Is it a problem you're hitting while using the browsable API? Is it that you client sends its payload with content-type: form-data instead of JSON? Something else?
> 
> It would be nice to get clarity on that, just to make sure there isn't a better to solve/workaround it in user-land.

That's actually very strong and valid argument you have put here @browniebroke .  I have been working with DRF from more than 3 yrs and I never came across this situation. But the point you made , if the "content type: form-data" , it create  scope for investigation. if it is breaking then it has be fixed but if not then without strong evident of failure of real world scenario  In my opinion , there is no  point in fixing it bcs it may break existing though i haven't tested it. 
I may be wrong but still we don't have any case (real world case) where it breaks. though all these are my observations and opinion.

**Comment by p-r-a-v-i-n** [comment]
This is just for visibility so far

```
class TestSerializer(serializers.Serializer):
    required = serializers.BooleanField(required=True)
    param = serializers.BooleanField(required=False)


@api_view(['POST'])
def test_api(request):
    try:
        serializer = TestSerializer(data=request.data)
        serializer.is_valid()
        return Response(serializer.validated_data)
    except Exception as e:
        print(str(e))
        return Response(str(e), 500)
```
Without changes in function
```
def get_value(self, dictionary):
        """
        Given the *incoming* primitive data, return the value for this field
        that should be validated and transformed to a native value.
        """
        if html.is_html_input(dictionary):
            # HTML forms will represent empty fields as '', and cannot
            # represent None or False values directly.
            if self.field_name not in dictionary:
                if getattr(self.root, 'partial', False):
                    return empty
                return self.default_empty_html
            ret = dictionary[self.field_name]
            if ret == '' and self.allow_null:
                # If the field is blank, and null is a valid value then
                # determine if we should use null instead.
                return '' if getattr(self, 'allow_blank', False) else None
            elif ret == '' and not self.required:
                # If the field is blank, and emptiness is valid then
                # determine if we should use emptiness instead.
                return '' if getattr(self, 'allow_blank', False) else empty
            return ret
        return dictionary.get(self.field_name, empty)
```
Api responses 
```
// request with form-data
curl -X POST http://localhost:8000/ -F "required=False"
// response
{"required":false,"param":false}

// request with application/json
 curl -X POST http://localhost:8000/ -H "Content-Type: application/json" -d '{"required": true}'
// response
{"required":true}
```

With changes
```
            if self.field_name not in dictionary:
                if getattr(self.root, 'partial', False):
                    return empty
                if not self.required:
                   return empty # with this changes 
                return self.default_empty_html
         
```
Api responses 
```
// request with form-data
curl -X POST http://localhost:8000/ -F "required=False"
// response
{"required":true}

// request with application/json
 curl -X POST http://localhost:8000/ -H "Content-Type: application/json" -d '{"required": true}'
// response
{"required":true}
```

**Comment by merwok** [comment]
```python
 def is_html_input(dictionary): 
     # MultiDict type datastructures are used to represent HTML form input, 
     # which may have more than one value for each key. 
     return hasattr(dictionary, 'getlist') 
```

Could this check be changed to look at the request content type? then it could differentiate between multipart/form-data and application/x-www-form-urlencoded, so that deserializing a query string can avoid the issue discussed here.

(You could mention that POST form data can use x-www-form-urlencoded too. But the browsable API doesn’t, and that’s what we want to identify here!)

**Comment by p-r-a-v-i-n** [comment]
make sense, relying on `hasattr(dictionary, 'getlist') `(i.e. checking for QueryDict) does not reliably distinguish between multipart/form-data, application/x-www-form-urlencoded, or query strings (request.GET).

If we can access the request object (e.g., via self.context['request']), 
we could modify is_html_input() to something like:
```
def is_html_input(request):
    return request.content_type == 'multipart/form-data'
```

This would make the behavior more precise and avoid false positives for url query strings or other cases where a QueryDict might appear.

@merwok Thanks.

**Comment by p-r-a-v-i-n** [comment]
> def is_html_input(dictionary): 
>      # MultiDict type datastructures are used to represent HTML form input, 
>      # which may have more than one value for each key. 
>      return hasattr(dictionary, 'getlist') 
> Could this check be changed to look at the request content type? then it could differentiate between multipart/form-data and application/x-www-form-urlencoded, so that deserializing a query string can avoid the issue discussed here.
> 
> (You could mention that POST form data can use x-www-form-urlencoded too. But the browsable API doesn’t, and that’s what we want to identify here!)

make sense, It would be more precise and robust way but we would require access to the request object there

**Comment by merwok** [comment]
Yes, the request would need to be passed from the serializer context to the method to the utils function.

**Comment by p-r-a-v-i-n** [comment]
> Yes, the request would need to be passed from the serializer context to the method to the utils function.

yes, but every time we have to explicitly pass the request object in context then
for example,
`serializer = TestSerializer(data=data, context={"request": request})`

are we open for this ?

**Comment by merwok** [comment]
I use serializers in viewsets 99% of the time, so they have a context with request
(edit: and I pass similar contexts in other kinds of views… but not in the one where I parse a query string! so worth noting in the docs if this gets changed).

IMO it would be fine if the is_html_input function had optional request param and returned False if not given!

**Comment by p-r-a-v-i-n** [comment]
> I use serializers in viewsets 99% of the time, so they have a context with request. IMO it would be fine if the is_html_input function had optional request param and returned False if not given!

In my opinion , It should be made mandatory then if ever serializer is instantiate.

**Comment by merwok** [comment]
🤷🏽 this is becoming off-topic. Serializers may or may not have a context with a request. DRF is stable now.

**Comment by browniebroke** [comment]
I'm yet to see a compelling real-world use case without workaround that would justify such a change, and until I see such description, I'd say that the project is unlikely to accept a fix for this, as the potential to break something else is too big.

Meanwhile, there is simple workaround in the original post that user can decide to use on a case-by-case basis.

[Also](https://www.django-rest-framework.org/community/contributing/)...

**Comment by browniebroke** [comment]
As discussed, not planning to fix this. The benefit/risk isn't worth it at this point.
